src/alerter.rs

```diff
@@ -0,0 +1,389 @@
+use crate::WatchtowerConfig;
+use crate::pagerduty::PagerDutyClient;
+
+use anyhow::Result;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, Instant};
+use tokio::sync::mpsc::{self, UnboundedSender, UnboundedReceiver};
+use tokio::sync::Mutex;
+
+#[derive(Deserialize, Clone, PartialEq, Eq, Debug, Default)]
+pub enum AlertLevel {
+    #[default]
+    None,
+    Info,
+    Warn,
+    Error,
+}
+
+#[derive(Clone, Debug)]
+pub struct AlertParams {
+    name: String,
+    description: String,
+    alert_level: AlertLevel,
+}
+
+impl AlertParams {
+    pub fn new(name: String, description: String, alert_level: AlertLevel) -> Self {
+        AlertParams { name, description, alert_level }
+    }
+
+    #[cfg(test)]
+    pub fn is_name_equal(&self, name: &str) -> bool {
+        self.name == name
+    }
+
+    #[cfg(test)]
+    pub fn is_description_equal(&self, description: &str) -> bool {
+        self.description == description
+    }
+
+    #[cfg(test)]
+    pub fn is_level_equal(&self, alert_level: AlertLevel) -> bool {
+        self.alert_level == alert_level
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct WatchtowerAlerter{
+    alert_sender: UnboundedSender<AlertParams>,
+    alert_receiver: Arc<Mutex<UnboundedReceiver<AlertParams>>>,
+    alert_cache: Arc<Mutex<HashMap<String, Instant>>>,
+    alert_cache_expiry: Duration,
+    pagerduty_client: Option<PagerDutyClient>,
+    watchtower_system_name: String,
+    allowed_alerting_start_time: SystemTime,
+}
+
+impl WatchtowerAlerter{
+    pub fn new(config: &WatchtowerConfig, pagerduty_client: Option<PagerDutyClient>) -> Result<Self> {
+        let alert_cache = Arc::new(Mutex::new(HashMap::with_capacity(config.alert_cache_size)));
```

Comment:
`HashMap::with_capacity` is not specifying a maximum size for the cache, but rather "an empty HashMap with **at least** the specified capacity."

In order to use this as a max size, we'd need to somehow pop the oldest values. Sounds a bit like a pain to be honest.

I'm a fan of using an existing library like [cached](https://docs.rs/cached/latest/cached/), specifically the [TimedSizedCache](https://docs.rs/cached/latest/cached/stores/struct.TimedSizedCache.html) where you could specify size and expiry time with `with_size_and_lifespan`. In this case the value you store doesn't have to be the time, as it will track the time itself under the hood.